### ISSUE TITLE
Add Recent Realtime LMP data to PJM

### ISSUE THREAD

**Issue by kmax12** [kmax12/gridstatus#103]
Currently, the PJM  implementation uses "Data Miner 2" for real-time LMP data in the `REAL_TIME_HOURLY` and `REAL_TIME_5_MIN` markets. However, I've observed a delay of approximately 3 days in the update of this data.

One possible solution is to look into getting the data from the "Data Viewer" application (https://dataviewer.pjm.com/dataviewer/pages/public/lmp.jsf) when the users requests data not yet available in Data Miner 2.

**Comment by kmax12** [comment]
another possible source: https://github.com/kmax12/gridstatus/issues/103

pjm is the only ISO without real time data available for LMPs, so high priority to add this

**Comment by jcsalterego** [comment]
> Currently, the PJM implementation uses "Data Miner 2" for real-time LMP data in the REAL_TIME_HOURLY and REAL_TIME_5_MIN markets.

I am not seeing this in the main branch - am I missing something?

At any rate, I started something in #132 but I'm not sure how to interpret some of the grid data.

**Comment by kmax12** [comment]
thank you so much for picking this up.

"data miner 2" is what get called in the `_get_pjm_json`. They call the application [data miner 2](https://dataminer2.pjm.com/list) but it ultimately just uses: https://api.pjm.com/api/v1/

I am away from my computer but will look at the PR this afternoon. I think I can help sort out some of the confusion

**Comment by kmax12** [comment]
so, i think we actually want to scrape the data from this chart since it is the time series for the last 3 days, which conveniently the data range what the other api doesn't have.

<img width="1219" alt="image" src="https://user-images.githubusercontent.com/371792/211171524-0579e7d6-ed9c-4c2e-b3e3-8c6877d532d1.png">

I think just the default locations that are selected work. does that chart seem like it would easy to scrape with the scaffolding you already have in place?

**Comment by jcsalterego** [comment]
Not sure, I'll have to take a closer look. Also I understand that it's missing from latest/today, so disregard my original question.

**Comment by kmax12** [comment]
actually, I think the day ahead market seems to have up to date data. 

both of these call work for me

```python
import gridstatus
import pandas as pd

iso = gridstatus.PJM()

today = pd.Timestamp("today", tz="US/Eastern")
one_day_ago = today - pd.Timedelta(days=1)
two_days_ago = today - pd.Timedelta(days=2)

iso.get_lmp(
    date=today,
    market=gridstatus.Markets.DAY_AHEAD_HOURLY,
    locations="hubs"
)

```

```python
iso.get_lmp(
    date=today+pd.Timedelta(days=1),
    market=gridstatus.Markets.DAY_AHEAD_HOURLY,
    locations="hubs"
)
```

The realtime markets seem to not have data though

```python
iso.get_lmp(
    date=one_day_ago,
    market=gridstatus.Markets.REAL_TIME_5_MIN,
    locations="hubs"
)
```
returns
```
RuntimeError                              Traceback (most recent call last)
/Users/kanter/Documents/isodata/docs/Examples/misc/pjm.ipynb Cell 2 in <cell line: 1>()
----> [1](vscode-notebook-cell:/Users/kanter/Documents/isodata/docs/Examples/misc/pjm.ipynb#W1sZmlsZQ%3D%3D?line=0) iso.get_lmp(
      [2](vscode-notebook-cell:/Users/kanter/Documents/isodata/docs/Examples/misc/pjm.ipynb#W1sZmlsZQ%3D%3D?line=1)     date=one_day_ago,
      [3](vscode-notebook-cell:/Users/kanter/Documents/isodata/docs/Examples/misc/pjm.ipynb#W1sZmlsZQ%3D%3D?line=2)     market=gridstatus.Markets.REAL_TIME_5_MIN,
      [4](vscode-notebook-cell:/Users/kanter/Documents/isodata/docs/Examples/misc/pjm.ipynb#W1sZmlsZQ%3D%3D?line=3)     locations="hubs"
      [5](vscode-notebook-cell:/Users/kanter/Documents/isodata/docs/Examples/misc/pjm.ipynb#W1sZmlsZQ%3D%3D?line=4) )

File ~/Documents/isodata/gridstatus/decorators.py:75, in support_date_range.__call__.<locals>.wrapped_f(*args, **kwargs)
     73 # no date range handling required
     74 if "end" not in args_dict:
---> 75     df = f(**args_dict)
     76     _handle_save_to(df, save_to, args_dict, f)
     77     return df

File ~/Documents/isodata/gridstatus/pjm.py:332, in PJM.get_lmp(self, date, market, end, locations, location_type, verbose)
    324 elif locations is not None:
    325     warnings.warn(
    326         (
    327             "Querying before archive date, so filtering by location will happen"
    328             " after all data is downloaded"
    329         ),
    330     )
--> 332 data = self._get_pjm_json(
...
--> 429     raise RuntimeError("No data found for query")
    431 df = pd.DataFrame(r["items"])
    433 num_pages = math.ceil(r["totalRows"] / row_count)

RuntimeError: No data found for query
```

Same error for hourly market after comment out t[he code](https://github.com/kmax12/gridstatus/blob/main/gridstatus/pjm.py#L266) disallowing it

The calls work for 2 days ago

**Comment by jcsalterego** [comment]
Yeah, Data Viewer uses https://www.oracle.com/java/technologies/javaserverfaces.html and appears to be all sorts of stateful. I can get the chart data but it requires a very specific dance. This will take a bit to do well but I think it's possible.

**Comment by kmax12** [comment]
is the latest section a lot easier to parse? if so, I think we just do that for since you seem to almost have it done.

**Comment by jcsalterego** [comment]
> is the latest section a lot easier to parse? if so, I think we just do that for since you seem to almost have it done.

It's about the same, I think. I'm close—but just adding a little bit of resiliency rather than hardcoding all the things. I've been chipping away at it so I'll have updated code in a few days or so.

If you're planning to uncomment the Market check and allow DAM then I can compare with the Data Viewer data and/or just use DV's RTM data.

**Comment by kmax12** [comment]
updated the pjm checks here: #133. is that what you had in mind? 

One other thing - I dont think there is actually a hourly real time market. it's only a 5 minute market. I'm guessing the "hourly" market that pjm [publishes on data miner](https://dataminer2.pjm.com/feed/rt_hrl_lmps/definition) is just an aggregation of 5 minute markets.

**Comment by jcsalterego** [comment]
> updated the pjm checks here: #133. is that what you had in mind?

Yep, cool beans.

> One other thing - I dont think there is actually a hourly real time market.

Noted

**Comment by jcsalterego** [comment]
![](https://user-images.githubusercontent.com/371792/211171524-0579e7d6-ed9c-4c2e-b3e3-8c6877d532d1.png)

#132 is updated to pull all the available locations (in `Select LMP(s)`) but it's a hodgepodge:

27 locations = ['AE', 'AEP', 'APS', 'ATSI', 'BATH COUNTY', 'BC', 'COMED', 'DEOK', 'DOM', 'DPL', 'DUQ', 'GALLATIN', 'GHENT_LG', 'IMO', 'JC', 'ME', 'MISO', 'PE', 'PEP', 'PJM-RTO', 'PL', 'PN', 'PS', 'RECO', 'WESTERN HUB', 'JABEZ', 'KLINE']

They appear to be mostly zones but `WESTERN HUB` shows up on the list as well.

What's the best way to incorporate this limited data set?

**Comment by kmax12** [comment]
Nice! A few notes

* The current status of the `PJM.get_lmp` is to return both a `Location` and `Location Name` column where the location is a pnode id from PJM, so we need to make this data consistent with the other methods, even if it returns fewer locations. 
* There is a `PJM.get_pnode_ids()` method that returns p nodes and their ids. I think this can be used to hydrate this data with location id and location type info. As things stand, it looks like 18/27 of these locations are direct match to what that return. The missing ones might also have matches, but haven't looked deeply yet

```python
iso = gridstatus.PJM()
df = iso._get_lmp_latest(verbose=True)
locations = df["Location"].unique().tolist()
pnodes = iso.get_pnode_ids()
matching_pnodes = pnodes[pnodes["pnode_name"].isin(locations)]
matching_pnodes
```
```
        pnode_id                pnode_name pnode_type pnode_subtype      zone voltage_level       effective_date     termination_date
0              1                   PJM-RTO  AGGREGATE          ZONE      None          None  2017-09-08T00:00:00  9999-12-31T00:00:00
1138       51288               WESTERN HUB  AGGREGATE           HUB      None          None  1998-04-01T00:00:00  9999-12-31T00:00:00
1141       51293                       DPL  AGGREGATE          ZONE      None          None  1998-04-01T00:00:00  9999-12-31T00:00:00
2074     7633629                      RECO  AGGREGATE          ZONE      None          None  2002-03-01T00:00:00  9999-12-31T00:00:00
2075     8394954                       APS  AGGREGATE          ZONE      None          None  1998-04-01T00:00:00  9999-12-31T00:00:00
2076     8445784                       AEP  AGGREGATE          ZONE      None          None  2004-10-01T00:00:00  9999-12-31T00:00:00
2265    31252687                       IMO  AGGREGATE     INTERFACE  EXTERNAL          None  2003-08-01T00:00:00  9999-12-31T00:00:00
3648    32411763        KLINE   138 KV  T1        BUS          LOAD       AEP        138 KV  2004-10-01T00:00:00  9999-12-31T00:00:00
4475    33092371                     COMED  AGGREGATE          ZONE      None          None  2004-05-01T00:00:00  9999-12-31T00:00:00
5464    34964545                       DOM  AGGREGATE          ZONE      None          None  2005-05-01T00:00:00  9999-12-31T00:00:00
5485    35010375               BATH COUNTY  AGGREGATE           EHV       DOM        500 KV  2005-05-01T00:00:00  9999-12-31T00:00:00
5534    37737283                       DUQ  AGGREGATE          ZONE      None          None  2005-01-01T00:00:00  9999-12-31T00:00:00
5926    40523629                      MISO  AGGREGATE     INTERFACE  EXTERNAL          None  2005-04-01T00:00:00  9999-12-31T00:00:00
7570   116013753                      ATSI  AGGREGATE          ZONE      None          None  2011-06-01T00:00:00  9999-12-31T00:00:00
8199   124076095                      DEOK  AGGREGATE          ZONE      None          None  2012-01-01T00:00:00  9999-12-31T00:00:00
8959  1123174808  GALLATIN138 KV  34.5KVLD        BUS          LOAD      EKPC        138 KV  2013-04-17T00:00:00  9999-12-31T00:00:00
8989  1123174840    JABEZ   161 KV  LD_143        BUS          LOAD      EKPC        161 KV  2013-04-17T00:00:00  9999-12-31T00:00:00
9148  1123175479    GHENT_LG345 KV  LD_122        BUS          LOAD      EKPC        345 KV  2013-04-17T00:00:00  9999-12-31T00:00:00
```
```python
# missing pnodes
>>> set(locations) - set(pnodes["pnode_name"].unique())
```

```
{'AE', 'ME', 'PE', 'BC', 'PN', 'PS', 'PEP', 'PL', 'JC'}
```

* One more point: the only other ISO that has Location Ids right now is ISONE and we actually do the opposite there, where there are `Location` (the name) and `Location Id` columns.  I'd like to flip PJM to return `Location` and `Location Id` since it is more consistent with the other ISOs and intuitive for user. However, the PJM API's deal with Ids, so need to think through the best way to do that. This should probably be done independently of your PR though.

**Comment by jcsalterego** [comment]
OK cool.

And I added
```python
        if re.search(r"[0-9]+ KV", item_id):
            # clean up Voltage and Station artifacts in name
            item_id = re.sub(r"([0-9]+) KV.*$", "", item_id).strip()
```

in a later commit... I'll take that out since it matches pricing nodes.

**Comment by kmax12** [comment]
resolved in #144
